### ISSUE TITLE
Can we discuss a better way to specify data context for component à la partials :)

### ISSUE THREAD

**Issue by JonDum** [ractivejs/ractive#2166]
I've yet to have a ractive project where I don't end up creating a loop of components like this:

```
 {{#each stuff}}
    <thing ../>
 {{/each}}
```

Where the  the `<thing/>` component needs to have several to dozens of properties set. It's asinine to map all of them as html `attribute="value"` pairs, _especially_ when they can change quite frequently when rapidly prototyping. 

Currently, I get around this by doing

```
 {{#each stuff}}
    <thing model="{{this}}"/>
 {{/each}}
```

and

```
thing = Ractive.extend({ 
  oninit: function() {

      /* Copy the model to the data  :|  :|  :|   */
      var model = this.get('model');
      this.set(model);

  }
})
```

But this workaround has been a super pain in the ass and it smells terrible. You get all sorts of oddities with two way binding and you have to be super careful not to overwrite properties, or even return new computed objects via slice or something—as any of that will break two way binding. It gets even more hairy when you realize that some primitives are always passed by value whereas objects aren't, so you have to wrap things in `{value: "foo"}` or instead revert to `<div>{{model.foo}}</div>` in your template.

What makes me most sad is that partials sort of already have this figured out and it makes components look second-class:

```
{{#stuff}}
    {{> thing someData}}
{{/}}
```

Would it be hard to do something like?

```
{{#stuff}}
    <thing someData/>
{{/}}
```

**Comment by Madgvox** [comment]
Perhaps a reserved `default` attribute that acts like `Ractive.defaults.data` for the component instance.

```html
{{#stuff}}
  <thing default="{{ someData }}" />
{{/}}

**Comment by martypdx** [comment]
Why not just reference the data implicitly and not pass anything? You can always pin down the parameters later...

**Comment by martypdx** [comment]
Also, see [this conversation](https://groups.google.com/forum/#!searchin/ractive-js/data$20components/ractive-js/5z1b-t1CsN0/pLRYRDLvNJkJ)

**Comment by Madgvox** [comment]
What @martypdx said. I forgot to mention the implicit data part

**Comment by JonDum** [comment]
@martypdx @Madgvox By 'implicit' do you mean accessing the parent's/roots properties directly as if they were on the component's viewmodel?

**Comment by Madgvox** [comment]
@JonDum yep, if you have `isolated:false` on the component, you can access the parent context.

**Comment by JonDum** [comment]
I see. What if you have components within components within components? What about naming conflicts? What about third party components via npm or web component? Implicitness makes it unclear what is part of the component and what is part of the parent chain, and tracking that down later leads to sloppier, more bug prone code.

To me, sharing parent contexts is a feature of a partial. When I need data encapsulation I go to a component, but currently this issue makes it very hard (I would even go as far as saying components should be isolated by default once we have some sort of mechanism to pass a data context).

I don't think it'd be that complicated of a feature to add, no? All I'm thinking is the `data: {}` object on the prototype would be defaults and merged with whatever is passed to the `<component data/>` call. Like a `set()` before the component is initialized. 

I wouldn't mind tackling it myself either if there's no resistance to that effort.

**Comment by Rich-Harris** [comment]
I tend to rely on implicit mappings a lot as well, but I can get away with that because my day job is to write code that doesn't require long term maintenance (as long as you can hammer it roughly into shape by the deadline, it doesn't really matter that it's covered in hammer dents). For larger projects, I agree that it's not ideal.

So I've occasionally done `<Thing model='{{this}}'/>` as well, but instead of `this.set(model)` (which causes all the problems you identified) I just have to remember to use `{{model.prop}}` and so forth. Also not ideal, but preferable out of those two options I think (you can always just wrap the entire component template in a `{{#with model}}...{{/with}}` if it's an issue).

Anyway: I quite like @Madgvox's suggestion of a reserved word, though I'd pick a different one:

```js
<Thing this='{{this}}' other='{{other}}'/>
```

`this` is already used in templates to mean 'the current thing', so that would make sense I think. Implementation-wise, I'd see this as working like so, if you found e.g. `{{foo}}` in the component template:

1. Does `foo` exist in the current context? If so use that
2. Does `foo` exist at the root of Thing's viewmodel, as data, a computed property or an explicit mapping? If so use that
3. Does the `this` object have a `foo` property? If so use that
4. Is Thing isolated? If so stop
5. Does `foo` exist in a parent context? If so use that
6. `foo` is unresolved

In other words, 3) is the only new step. It *wouldn't* be 'mixed in' with Thing's data, it would be another form of mapping (this is more efficient, preserves SSOT and all that).

**Comment by fskreuz** [comment]
> So I've occasionally done <Thing model='{{this}}'/> as well, but instead of this.set(model) (which causes all the problems you identified) I just have to remember to use {{model.prop}} and so forth.

This is what I've been doing in components and has worked pretty well so far.

```html
<!-- usage -->
<time-widget time="{{ timeObject }}" />

<!-- definition -->
<div class="time-widget">
  <span>{{time.hours}}</span>:
  <span>{{time.minutes}}</span>:
  <span>{{time.seconds}}</span>
</div>
```

The only hurdle I face is to remember that `time` is an object. And being an object, the default value must be wrapped in a data function. Other than that, everything Ractive has already covers this.

> Anyway: I quite like @Madgvox's suggestion of a reserved word, though I'd pick a different one:

`data` would be a good property name to be consistent with `data` in the config. This would auto-spread the object into the top-level properties. For instance, if `<time-widget>` accepted `hours`, `minutes` and `seconds`, the following would be equivalent:

```html
<!-- timeObject : {hours: 11, minutes: 59: seconds: 42 } -->

<time-widget data="{{ timeObject }}" /> 

<time-widget hours="{{ timeObject.hours}}" minutes="{{ timeObject.minutes }}" seconds="{{ timeObject.seconds }}" />
```

If this spread is done, and a property is explicitly defined as well, I'd vote that the explicit one override the one from the object.

my 2c.

**Comment by Rich-Harris** [comment]
I'm a little nervous about `data` because it's not wildly uncommon for an object to have a property by that name, in my experience. Whereas `this` is much less likely, partly because it's a keyword but also because e.g. `user.this` doesn't really make any sense whereas `user.data` totally does. Also, in a template [you can do](http://jsfiddle.net/vL3203g3/) `{{this.foo}}` (you wouldn't, because you'd just use `{{foo}}`, but it basically means the same thing), so the semantics dovetail nicely.

A separate point to consider though: with `<Thing this='{{myObj}}'/>`, if inside Thing we set data that *wasn't* on `myObj` (e.g. `this.set('foo', 'bar')`, when there's no `myObj.foo`), does that mutate `myObj`? My feeling is that it shouldn't, but we should probably think it through.

**Comment by heavyk** [comment]
+1 on keyword `this`. I use `<Thing d="{{.}}" />` a lot. having `<Thing this="{{.}}" />` would be totally awesome

+1 +1 :)

**Comment by Madgvox** [comment]
+1 for `this`, much clearer intent.

**Comment by JonDum** [comment]
I like the `this='{{this}}'` syntax as well and the extra step `3` would totally solve my problem! If I'm to play devil's advocate though, I'd say that it looks like a normal and valid html attribute so the new user might not think it does anything special until they read a small paragraph in the Components docs. `<Thing {{this}} other="{{other}}"/>`, on the other hand, is clear that something special is going on and if the user has any familiarity with partials will recognize the similar syntax, no?

----

&nbsp;

> A separate point to consider though: with <Thing this='{{myObj}}'/>, if inside Thing we set data that wasn't on myObj (e.g. this.set('foo', 'bar'), when there's no myObj.foo), does that mutate myObj?

Definitely yes, which is why I suggested just `set()`ing the data passed in instead of maintaining another "ghost" `this` object that needs to be tracked and updated. 

I've made two major SPAs with Ractive now and I use components to encapsulate everything from entire pages of the site to autosizing text inputs, to complex "blocks" in a [block programming application](http://j2inn.com/playground/block-programming). Most of the time I would not only want `myObj` to be updated, but every *new*, non-computed property on the `data` viewmodel of the component should be synced to the backing parent context (`myObj`, in your scenario). This is actually how I thought isolated worked at first!

I assumed that was the case because of the way partials worked. The docs talk heavily about how partials share context but can also be passed a specific context, Yet, if you need data encapsulation you should use a component. At the time, I thought that it would be silly for components to be able to access parent references like that because that's what partials did so *surely* `isolated` did something akin to not updating the backing context `myObj` instead of just blocking parent references (which isn't that useful of a thing imo).

**Comment by martypdx** [comment]
While I understand the desire, I suspect the dichotomy between `data` as registry of data (which it currently is) versus `data` as an meaningful domain object (which this would make it) means it will not end by the time all the upwards updating cases are considered.

As a alternative suggestion, what if ractive options included a `defaultContext` that accepted a keypath and essentially used that as the starting context for the template?

```js
const Time = Ractive.extend({
    defaultContext: 'time',
    ...
}); 
```

```html
<!-- usage -->
<time-widget time="{{ timeObject }}" />

<!-- definition -->
<!-- implicit {{#with time}} -->
<div class="time-widget">
  <span>{{hours}}</span>:
  <span>{{minutes}}</span>:
  <span>{{seconds}}</span>
</div>
```

That's what made partial context work, would be more broadly useful, and tons simpler to implement.

**Comment by martypdx** [comment]
@JonDum my suggestion to use implicit data was directed at your "especially when they can change quite frequently when rapidly prototyping." comment, not that you should or shouldn't use `isolated` one way or the other.

**Comment by JonDum** [comment]
> my suggestion to use implicit data was directed at your "especially when they can change quite frequently when rapidly prototyping." comment, not that you should or shouldn't use isolated one way or the other.

@martypdx Ah for sure. I think it's a good suggestion but I hadn't correctly communicated my real reasoning at that point (it was also like 3am :P).

**Comment by JonDum** [comment]
>As a alternative suggestion, what if ractive options included a defaultContext that accepted a keypath and essentially used that as the starting context for the template?

That's an interesting idea actually. I presume it would be like wrapping the template in a `{{#with}}` statement for you?

**Comment by Rich-Harris** [comment]
> Most of the time I would not only want `myObj` to be updated, but every new, non-computed property on the data viewmodel of the component should be synced to the backing parent context

I had a bit of a think about this, and it occurred to me that there's a very good reason you *wouldn't* want that behaviour – a component could never have private properties! It would be like destroying the distinction between `this.props` and `this.state` in a React component. For example if a component had a highlight effect on hover, you probably wouldn't want that reflected higher up – it would be a private `highlighted` property or something. Whereas going the other way, you could always force a link between `foo` (in the component) and `myObj.foo` by ensuring that `myObj` had a `foo` property (`null`, if necessary) to start with.

> what if ractive options included a `defaultContext` that accepted a keypath

Would that just apply to the template? Or would `this.get('minutes')` with a `defaultContext` of `time` be the same as `this.get('time.minutes')` without one?

**Comment by JonDum** [comment]
> a component could never have private properties

Yea this is tricky. I personally don't mind if a var gets reflected higher up, because it'll be encapsulated within the object that it was passed anyways — it doesn't leak to the parent, just an object on the parent. 
I can't say the same for everyone else, though.

**Comment by martypdx** [comment]
Oh, I finally got what @Rich-Harris meant: instead of trying to set the data, set the context. The implicit reference lookup functionality works correctly, it's just overly broad in what it allows.

Yeah, that makes a lot of sense. Avoids the messy redefining of root and would even support both `isolated` settings. Would totally work, we should do that :)

Would it be more clear if we used `@context`? Or is the not strictly legal attribute an issue?

```html
{{#each stuff}}
    <thing @context='{{this}}' foo='{{bar + 2}}' index='{{@index}}' on-qux='something()'/>
{{/each}}
```
It should also override `event.context` for component events.

**Comment by thomsbg** [comment]
Throwing in my 2¢:

For context, I use ractive to build a multi-page app, and I set `Ractive.defaults.isolated = true` for maintainability reasons.

Adding a new step in the resolution algorithm (as proposed by @Rich-Harris) seems like heaping complexity on top of what is already a pretty tricky concept. `data`, computed properties, explicit mappings, AND a `this` mapping all contributing to a "load path" for template variables seems like too many things IMO. Merging multiple sources of data (with the potential for name conflicts & undefined behavior) gets more dangerous as the number of sources goes up. Perhaps it's growing pains with using Ractive in a larger app, but if anything, I would prefer to explicitly separate these sources of data somehow, similar to `props` vs `state`.

I'm not a big fan of the JSX syntax, but I do appreciate how `<Component {...this.props} plus="extra">` leans on ES6 rest parameters / destructuring paradigms. In Ractive, maybe it could look like: `<thing {{...this}} plus="more">`. To me, this reads as setting up individual explicit links for each own property of `this`, which is more grok-able than a new resolution step.

**Comment by JonDum** [comment]
@thomsbg I'm glad I'm not the only one who prefers that syntax. 

Also, the current solution rich/chris has proposed doesn't merge anything like I suggested. It's more like a couple series of checks. Is it defined on the component's data? No? Ok, now check the alternate context. Still not defined? Ok check it's parent if it's not isolated... etc.

**Comment by thomsbg** [comment]
Maybe "merge" was the wrong word. After dwelling on it, the main thing I worry with adding those extra checks is ambiguity about what to do for `set()`, or what would happen when keys are added or removed from the provided context. For example:

```html
// inside outer-component, myObj is { foo: 'bar' }
<inner-component this="{{ myObj }}">
```

1. What happens when outer-component does `set('myObj.newKey', 'something')`?
1. What happens when outer-component does `set('myObj', 1234)`?
1. What happens when outer-component does `set('myObj', { totallyDifferentKey: 'value' })`?
1. What happens when inner-component does `reset({ something: 'else' })`?
1. What happens when inner-component does `set('newKey', 'something')?

Having a component's data context act dynamically like this makes me sweat. What if instead of dynamically passing along each key/value pair from an object, there was a shorthand for sending along named keys from an object, a la ES6 shorthand `{ foo, bar } = myObj`? Something like:

```html
<inner-component foo,bar,baz="{{myObj}}">

// or perhaps
{{#myObj}}
  <inner-component {{foo,bar,baz}}>
{{/myObj}}

// simply having a shorthand for foo="{{foo}}" would be helpful
<inner-component {{=foo}}>
```

Not married to any of these ideas, just brainstorming here.

**Comment by martypdx** [comment]
@thomsbg:

> What happens when outer-component does set('myObj.newKey', 'something')?

Totally depends on whether or not inner-component uses `{{newKey}}`. If not, nothing would happen (inside inner-component).

> What happens when outer-component does set('myObj', 1234)?

In your example assuming `{{foo}}` existed in inner-component template, it would now be undefined.

> What happens when outer-component does set('myObj', { totallyDifferentKey: 'value' })?

Again, `{{foo}}` would be undefined. If the inner-component references `{{totallyDifferentKey}}`, it would now have a value of `value`.

> What happens when inner-component does reset({ something: 'else' })?

Good edge case, we'll have to see. I suspect what your doing is resetting the default component data.

> What happens when inner-component does set('newKey', 'something')?

If the inner-component does not have data defined as `newKey`, then it would be set on `myObj` as `myObj.newKey`. 

This makes sense when you think about components as not having control over the context being set, but having complete control what properties will be utilized from the context and what default data should be. 

Consider `<user this='{{user}}'/>` with component template: `First: <input value='{{first}}'>, Last: <input value='{{last}}'>`. 

It doesn't matter if it gets passed `{ first: 'Marty', last: 'Nelson' }`, `{ last: 'Nelson' }`, `{}`, or even `undefined` - all work correctly.

Now if I want to define component behavior (state) versus data, I can add to the component definition `data: { readonly: false }` and change template to ` <input value='{{first}}' readonly='{{readonly}}'>`, etc. and now this property will never be added to or read from the context. But it can be provided as an additional parameter: `<user this='{{user}}' readonly='true'/>`. (IMO will still could benefit from separate state bucket from data bucket, but that's for another day)

**Comment by JonDum** [comment]
Closing since `this="{{this}} "` is implemented in edge :+1:

**Comment by evs-chris** [comment]
Is it?

**Comment by JonDum** [comment]
Yes? http://jsfiddle.net/n1u0cf85/

**Comment by evs-chris** [comment]
Ah! Unfortunately no. If you turn on isolation, you lose your ability to resolve from template context and the fiddle stops working: http://jsfiddle.net/n1u0cf85/1/

For a minute there, I thought I spaced on how that would have had to have been implemented. :smile:

**Comment by JonDum** [comment]
Oh yea, doh! I always have isolation turned on, so I forgot that it's off by default.

**Comment by evs-chris** [comment]
You can achieve a [mapping spread with a partial](https://ractive.js.org/playground/#N4IgFiBcoE5SAbAhgFwKYGcUgL4BoRtIQAeAEwEsA3AAgrIF4AdQpGAczRRYD4SB6SlR4gCGeFTY0YNBjQB2aAO40ASkgDGKamgAUwJvKYoUbTikg0A5AGJTHLlbyHj6ALYAHZOktXgwGzRNMDp3DBwcEg1PGn8eAAEMMDY0MgA6DA8YILJdNIBKCJp+Hn9+II0wCKcXFAB7eUsDI25teQoUXXzY2uNXMAoMDK5dK0Tk7PTM7KQyJxoAMwBXeS0KBt06gCMAK27mvtbjDQaMOoQ0NIQ69k3d-N6jlGyUJZh5WJoLGgB5XbQtGkANZoACeGDuezSbiQHl0QNkPBo+i+lgAjABmPAKSxA7ELSAAbWAqJoACZsTBcTQcABdGn5bo4ADcj2MOHyrJa7Nq+FqJ08DTQ8hQGCaj2iHks6jWVEuaAAHuh5LkDk93F5UGhfORqKVgAs6nUiv4tmwTcAzQAvCICIQ8KyPDm8wwcrmGDBcAAqFDcaDqS06y1W2g2+1qMDSHiWSVGHTQbgw8xJhrq6IpNDNVOs6CwVgZXPwNAArAAGUuclzyT0oH1+gNBlZrMM9FqRmtxsJpUtpa3zAAsZMr8iLaLLFfd1e9vv9gd0web8i6reMkejsas8cTyczSCtlgAnAeC67sWjyxPcEA) now, which isn't exactly setting the context, but I think between that sort of thing and components where a context name falls out naturally, this is covered to close enough. If not, feel free to reopen 😄.
